### PR TITLE
Refactor release workflow to be reusable

### LINKED ISSUE #212: [Release] Refactor to use reusable workflows
Whilst trying to do the v0.2.0 release there appears to be some areas of simplicication:

- We don't need to build the images twice, we can just tag them
- ghcr-provenance isn't used by anything.

### PR DESCRIPTION
**Pull request by Danil-Grigorev**
**What this PR does / why we need it**:

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #212

**Special notes for your reviewer**:

**Checklist**:

- [ ] squashed commits into logical changes
- [ ] includes documentation
- [ ] adds unit tests
- [ ] adds or updates e2e tests

### CODE REVIEW

**Review by salasberryfin — APPROVED** [review]
LGTM. Based on these changes, I have published a PR to address the release testing which depends on this one #235

File: .github/workflows/release_build/action.yaml [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+      uses: actions/setup-go@v4
+      with:
+        go-version: '=1.20.7'
+    - name: Docker login to ghcr registry
+      uses: docker/login-action@v3
+      with:
+        registry: ${{ inputs.registry }}
+        username: ${{ inputs.username }}
+        password: ${{ inputs.password }}
+    - name: Build docker image for gh registry
```
Comment (richardcase) [inline .github/workflows/release_build/action.yaml:50]:
```suggestion
    - name: Build docker image
```

File: .github/workflows/release_build/action.yaml [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+      shell: bash
+      run: make docker-build-${{ inputs.arch }} TAG=${{ inputs.tag }} REGISTRY=${{ inputs.registry }} ORG=${{ inputs.org }}
+    - name: Push docker image to gh registry
```
Comment (richardcase) [inline .github/workflows/release_build/action.yaml:53]:
```suggestion
    - name: Push docker image to registry
```

File: .github/workflows/release_sign/action.yaml [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+    type: string
+  password:
+    required: true
+    description: 'The password key to fetch from secret store for registry login'
+    type: string
+
+runs:
+  using: 'composite'
+  steps:
+  - name: Docker login to ghcr registry
```
Comment (richardcase) [inline .github/workflows/release_sign/action.yaml:40]:
```suggestion
  - name: Docker login to registry
```

**Review by richardcase — COMMENTED** [review]
This looks good. What does this mean in relation to PR #223?

**Comment by Danil-Grigorev** [comment]
This refactor includes signing images using digest instead of a tag @richardcase

File: .github/workflows/release.yaml [hunk 4]
```diff
@@ -1,205 +1,54 @@
...
-          image=$(echo ${{ matrix.images.image }} | base64 -d | base64 -d)
-          cosign verify ${image} --certificate-identity=https://github.com/rancher-sandbox/rancher-turtles/.github/workflows/release.yaml@refs/tags/${{ env.TAG }} --certificate-oidc-issuer=https://token.actions.githubusercontent.com
+        destination: [ghcr, prod]
+        arch: [amd64, arm64, s390x]
+        include:
+        - destination: ghcr
+          registry: ghcr.io
+          username: ${{ github.actor }}
+          image: ghcr.io/rancher-sandbox/rancher-turtles
+          password: GITHUB_TOKEN
```
Comment (richardcase) [inline .github/workflows/release.yaml:26]: Does this not need to be:

```suggestion
          password: ${{ secrets.GITHUB_TOKEN }}
```

Or am i misunderstanding this?

Comment (Danil-Grigorev) [reply .github/workflows/release.yaml:26]: Yes, we are passing keys for the secrets, not the secrets itself. Each job individually fetches the password, with the permissions assigned to it by `secrets[inputs.password]`

File: .github/workflows/release-workflow.yml [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+        with:
+          image: ${{ inputs.secret_registry && secrets[inputs.image] || inputs.image }}-${{ inputs.arch }}
+          digest: ${{ needs.build.outputs.digest }}
+          identity: https://github.com/${{ inputs.org }}/rancher-turtles/.github/workflows/release.yaml@refs/tags/${{ inputs.tag }}
+          oids-issuer: https://token.actions.githubusercontent.com
+          registry: ${{ inputs.secret_registry && secrets[inputs.registry] || inputs.registry }}
+          username: ${{ inputs.secret_registry && secrets[inputs.username] || inputs.username }}
+          password: ${{ secrets[inputs.password] }}
+
+  # Changes from https://github.com/slsa-framework/slsa-github-generator/pull/2918 are required to make the provenance work for production registry
```
Comment (furkatgofurov7) [inline .github/workflows/release-workflow.yml:94]: Referred changes seem to be merged, is the plan to enable the provenance not in this PR but later?

**Review by furkatgofurov7 — APPROVED** [review]

File: .github/workflows/release-workflow.yml (hunk 5 above)
Comment (Danil-Grigorev) [reply .github/workflows/release-workflow.yml]: Yes, it was merged but was not released. I’ve switched the image to be a collected from vars as was [recommended](https://github.com/slsa-framework/slsa-github-generator/issues/2917#issuecomment-1783508999), as it is not a concern to leave image name in logs for us, and recor transparency log will still contain the full image name.

Comment (richardcase) [reply .github/workflows/release-workflow.yml]: We'll leave the provenance commented out until we fully decide what we are going to do.

Comment (richardcase) [reply .github/workflows/release-workflow.yml]: Which remind me i need to create an issue

**Review by richardcase — DISMISSED** [review]

File: .github/workflows/release_sign/action.yaml [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+runs:
+  using: "composite"
+  steps:
+    - name: Docker login to registry
+      uses: docker/login-action@v3
+      with:
+        registry: ${{ inputs.registry }}
+        username: ${{ inputs.username }}
+        password: ${{ inputs.password }}
+    - uses: sigstore/cosign-installer@v3.1.2
```
Comment (richardcase) [inline .github/workflows/release_sign/action.yaml:46]: There appears to be an issue with the version its trying to install from looking at the actions:

```
INFO: Downloading bootstrap version 'v2.0.0' of cosign to verify version to be installed...
      https://storage.googleapis.com/cosign-releases/v2.0.0/cosign-linux-amd64
ERROR: Unable to validate cosign version: 'v2.0.0'
Error: Process completed with exit code 1.
```

Shall we force it to use a specific version in our workflow? So something like:

```
uses: sigstore/cosign-installer@v3.1.2
with:
  cosign-release: 'v2.2.0'
```

Comment (Danil-Grigorev) [reply .github/workflows/release_sign/action.yaml:46]: This is unrelated to the PR: https://github.com/rancher-sandbox/rancher-turtles/pull/247

Comment (richardcase) [reply .github/workflows/release_sign/action.yaml:46]: Ah yes.

**Comment by richardcase** [comment]
#247 has merged so we should be able to make the checks go green :)

**Review by richardcase — DISMISSED** [review]

File: .github/workflows/release-workflow.yml (hunk 5 above)
Comment (furkatgofurov7) [reply .github/workflows/release-workflow.yml]: > We'll leave the provenance commented out until we fully decide what we are going to do.
Looks like we enabled it in the same PR..

**Review by furkatgofurov7 — APPROVED** [review]

**Review by salasberryfin — CHANGES_REQUESTED** [review]

File: .github/workflows/release-workflow.yml (hunk 5 above)
Comment (richardcase) [reply .github/workflows/release-workflow.yml]: Yes, it seems Danil uncommented provenance after this comment.

File: .github/workflows/release.yaml [hunk 7]
```diff
@@ -1,205 +1,48 @@
...
+      REGISTRY_PASSWORD: ${{ secrets.REGISTRY_PASSWORD }}
 
   release:
     name: Create helm release
-    needs: [prod-sign]
+    needs: [build-push-services]
     runs-on: ubuntu-latest
     env:
       TAG: ${{ github.ref_name }}
-      PROD_REGISTRY: ${{ secrets.REGISTRY_ENDPOINT }}
```
[the remark this answers is not in the document]
Comment (Danil-Grigorev) [reply .github/workflows/release.yaml]: Fixed

**Review by salasberryfin — APPROVED** [review]
Thanks @Danil-Grigorev

**Review by richardcase — APPROVED** [review]

File: .github/workflows/release-workflow.yml (hunk 5 above)
Comment (Danil-Grigorev) [reply .github/workflows/release-workflow.yml]: Yes, the data is stored in the variable, and is passed to the job, avoiding scenarios when the job output is redacted.

**Review by furkatgofurov7 — APPROVED** [review]

File: .github/workflows/release-workflow.yml (hunk 5 above)
Comment (richardcase) [reply .github/workflows/release-workflow.yml]: @Danil-Grigorev @furkatgofurov7 - can we resolve this conversation? So this PR can be merged.

**Pull request merged**